rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,400 @@
+  be upscaled even in standalone mode. Deploying those nodes using a Deployment
+  means we cannot mount peristent storage on them. As Teleport does not support
+  graceful shutdown with record shipping, users might loose active sessions
+  recordings during a rollout if using the `proxy` mode. Teleport nodes
+  are relying on `kube` ProvisionTokens to join the auth nodes on startup ([see
+  RFD-0094](https://github.com/gravitational/teleport/blob/rfd/0096-helm-chart-revamp/rfd/0096-helm-chart-revamp.md)).
+- the `teleport-auth` Deployment:  Those pods cannot be
+  replicated without remote backend for state and audit logs. When persistence is
+  enabled, a single volume will be mounted to those pods and the update strategy
+  will be "re-create". For setups in which auth pods are stateless, the Deployment
+  can be scaled up.
+
+The main LB service should send traffic to the proxies, two additional services
+for in-cluster communication should be created: one for the proxies and one for
+the auth.
+
+The trust between auth and proxy should be bootstrapped by
+[creating a provisionToken on start](https://github.com/gravitational/teleport/pull/19009).
+
+#### Labels and selectors
+
+Deploying different pod sets requires a way to discriminate them. The only
+label set currently is `app: {{.Release.Name }}`. We should follow [Helm
+label recommendations](https://helm.sh/docs/chart_best_practices/labels/):
+
+| Label                        | Value                                 | Purpose   |
+|------------------------------|---------------------------------------|-----------|
+| app.kubernetes.io/name       | `{{- default .Chart.Name .Values.nameOverride \| trunc 63 \| trimSuffix "-" }}` | Identify the application. |
+| helm.sh/chart                | `{{ .Chart.Name }}-{{ .Chart.Version \| replace "+" "_" }}` | This should be the chart name and version. |
+| app.kubernetes.io/managed-by | `{{ .Release.Service }}`              | It is for finding all things managed by Helm. |
+| app.kubernetes.io/instance   | `{{ .Release.Name }}`                 | It aids in differentiating between different instances of the same application. |
+| app.kubernetes.io/version    | `{{ .Chart.AppVersion }}`             | The version of the app. |
+| app.kubernetes.io/component  | `auth` or `proxy`                     | This is a common label for marking the different roles that pieces may play in an application. |
+
+Those labels should be applied to all deployed resources when applicable.
+This includes but does not limit to Pods, Deployments, ConfigMaps,
+Secrets and Services.
+
+The `app: {{.Release.Name}}` label should stay on the auth pods for
+compatibility reasons.
+
+#### Monitoring
+
+A single optional
+[`ServiceMonitor`](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#servicemonitor)
+should be deployed per Helm release, selecting all release services based
+on `app.kubernetes.io/name`.
+
+#### Custom Resources
+
```

Comment:
Since helm can't deploy the CRD and its CRs in the same release, would it be possible to support bootstrap.yaml in the helm chart? That would hopefully allow folks to do a one-shot initial deploy that doesn't require them to separately create custom resources or exec into the pod.

Reply:
Technically you can do `bootstrap.yaml` with the Helm chart currently using `extraVolumes`, `extraVolumeMounts`  and `extraArgs`. Maybe if we just document how to do this better as it's intended to be a one-off thing?

Reply:
extraVolume/extraVolumeMounts/extraArgs also require that the end-user do a two-step operation to get that initial data in place. Allowing the chart to natively support as much of this initialization process as possible all from the chart values will reduce the need for external steps/scripts/processes.

Reply:
From what I read, `--bootstrap` does not support all resources (User, CA, TrustedCluster, Role, AuthConnectors) and allows you to import severely broken states (it does some validation about the CA material but does not check if a user has existing roles for example). Therefore, the flag should be used only for restoring an existing state and must not be used as a substitution for the CR/CRD mechanism.

We do need to make populating resources easier, but I think `--boostrap` is not the adequate way.